src/Microsoft.OData.Edm/Schema/EdmStructuredType.cs

```diff
@@ -163,6 +164,20 @@ public EdmNavigationProperty AddUnidirectionalNavigation(EdmNavigationPropertyIn
             return property;
         }
 
+        /// <summary>
+        /// Creates and adds a structural property alias to this type.
+        /// </summary>
+        /// <param name="declaring">Name of the property.</param>
+        /// <param name="type">Type of the property.</param>
+        /// <returns>Created structural property.</returns>
+        public EdmStructuralPropertyAlias AddStructuralPropertyAlias(IEdmTypeReference type, string alias, IEnumerable<string> path)
+        {
+            string propertyName = path.FirstOrDefault();
+            EdmStructuralPropertyAlias property = new EdmStructuralPropertyAlias(this, propertyName, type, alias, path);

```

Comment:
We should probably confirm that `this` type does indeed have properties that follow `path` and result in a property with name `propertyName` and of type `type`.

Reply:
Am not sure I understand this. So, here we are adding a property - `propertyName` of type `type` to `this` type

Reply:
I'm sure most of my comments around this has been nonsensical regarding what an alias means, but I think 
my points stands. Hopefully this example will help make sure I'm understanding correctly. There is a type `buzz` with a property `foo/bar/fizz` that is of type `Edm.Decimal`. The caller wishes to add an alias to `buzz` called `frob` that points to `foo/bar/fizz`. 

In this case, doesn't `frob` *have* to be of type `Edm.Decimal`? If that's the case, then what I'm suggesting is that we need to walk the `path` parameter in this method to ensure that:
1. Each segment of the path is in fact valid for `this`
2. The type of the property pointed to by `path` must match `type`

Reply:
The `key` for the type `buzz` is not in `buzz`. The key is in another type let's `fizz`. But the type `buzz` has a complex property `bar` whose type is `fizz` and `fizz` has a property `Id` whose type is `Edm.Int`. This is a 2-segment path. we could have a longer segment than that but that should be the flow. So the key for `buzz` will be represented as `bar/Id`.  The alias is an identifier for this path `bar/Id`. Instead of using this path, we use an alias which does not have the slashes and that is easy to work with. The Alias will always be a string.

Reply:
so the method `AddStructuralPropertyAlias` is used to add `bar` to `buzz` but because this property `bar` acts a link to the actual `key` property for type `buzz`, that's why we provide more information to it like the alias and the key path. The path is later processed to get to the last path segment to get the real `key` property for the type.

Reply:
```
var model = new EdmModel();
var entity = new EdmEntityType("NS1", "Product");
var complexType = new EdmComplexType("NS1", "ExampleComplexObject");
complexType.AddStructuralProperty("Name", EdmCoreModel.Instance.GetString(false));
var keyProperty = entity.AddStructuralPropertyAlias(new EdmComplexTypeReference(complexType, isNullable: false), "ExampleObjectName", new List<string> { "ExampleObject", "Name" });
entity.AddKeys(keyProperty);
entity.AddStructuralProperty("Name", EdmCoreModel.Instance.GetString(false));
```

In this test that you wrote, for this line: `var keyProperty = entity.AddStructuralPropertyAlias(new EdmComplexTypeReference(complexType, isNullable: false), "ExampleObjectName", new List<string> { "ExampleObject", "Name" });`, doesn't this *imply* that the `Product` entity has a property named `ExampleObject` of type `ExampleComplexObject` and that `ExampleComplexObject` has a property named `Name`? There are two things that are weird to me:
1. Can't I also call `var keyProperty = entity.AddStructuralPropertyAlias(new EdmComplexTypeReference(complexType, isNullable: false), "ExampleObjectName", new List<string> { "ExampleObject",  "*Garrett*" });`, even though that would be incorrect (because there is no property named "*Garrett*" on `ExampleComplexObject`)?
2. Why is it that the `ExampleObject` property of type `ExampleComplexObject` is added implicitly to `Product` when this method is called? Shouldn't I need to call `entity.AddStructuralProperty("ExampleObject", complexType)` first?

Reply:
@corranrogue9 
1. For the first case, Yes the user can pass   this ` "*Garrett*" ` that does not exist in `"ExampleObjectName"` but this will lead to model validation errors and an exception will be thrown. 
2. For the second case, If you want to support key alias in your model, then you use the `entity.AddStructuralPropertyAlias(...)` method  instead of using `entity.AddStructuralProperty(..)`

Reply:
Number 2 makes sense now, thanks for explaining that. 

For number 1, my point is, when we internally instantiate `EdmStructuralPropertyAlias`, we only do so through the use of `PropertyAliasHelpers.GetKeyProperty`, which guarantees the correctness of the alias in the model. If we moved the logic in that method to the constructor of `EdmStructuralPropertyAlias`, we would guarantee the same correctness, but for *non-internal* callers. It was *also* allow us to have access to both the original path and the split path segments (since they could be computed by the constructor), so we wouldn't need to call `string.Join` on the path in other parts of the code.

Reply:
For performance, open/close principle, and single-responsibility reasons, even if the model validation catches it, the logic should move to the constructor.

Reply:
@corranrogue9 I had @xuzhg mention that by design EDM does not validate the model unless the user explicitly chooses to do so. But maybe the context was different.

Reply:
This logic is used in computing the declared key. I don't know whether it can move to the constructor.

Reply:
@corranrogue9 So when we are adding the `complex property` to the model, we call `AddStructuralPropertyAlias`. Within this method we call `EdmStructuralPropertyAlias`. We can move the logic for checking the path and everything to the constructor. This will check for validity of the path. When we call `AddKeys` method for adding the key to the model, if we say that we pass in the last segment on the path and not do any validation because we already did it in the constructor, I think this will open up some loopholes where someone can just pass in something that is not even part of the path. That's why when we are adding the keys to the entity type's `DeclaredKey`, we still have to go through the path and make sure that whatever we are adding to the `DeclaredKey` is in the path and is the last segment on the path.